doc/README.md

```diff
@@ -0,0 +1,143 @@
+for that type are value-stable where expected to be.
+
+Before 1.0, we allow any new *minor* version to break value-stability, though
+we do expect such changes to be mentioned in the changelog. Post 1.0 we have
+not yet determined exact stability rules.
+
+Additionally, we expect patch versions not to change the output of any
+deterministic functions, even if not value-stable (this is not a hard
+requirement, but exceptions should be noted in the changelog).
+
+Defining which parts of Rand are value-stable is still in progress. Many parts
+of `rand_core` have some documentation on value-stability.
+
+### Project Scope
+
+The `rand_core` library has the following scope:
+
+-   the core traits which RNGs may implement
+-   tools for implementing these traits
+
+The `rand` library has the following scope:
+
+-   re-export all parts of `rand_core` applicable to end users
+-   an interface to request entropy from an external source
+-   hooks to provide entropy from several platform-specific sources
+-   traits covering common RNG functionality
+-   some PRNGs, notably `StdRng` and `SmallRng`
+-   `thread_rng` auto-seeding source of randomness
+-   conversion of random bits to common types and uses
+-   shuffling and sampling from sequences
+-   sampling from various random number distributions
+
+Note: the scope of the project and above libraries may change. We are currently
+discussing moving PRNGs (#431) and distributions (#290) to other libraries or
+projects.
+
+### Code style
+
+We do not currently have many policies on style other than:
+
+- neat and consistent
+- minimise changes which are purely stylistic, or move to a separate commit
+
+Rand does **make use of `unsafe`**, both for performance and out of necessity.
+We consider this acceptable so long as correctness is easy to verify.
+In order to make this as simple as possible,
+we prefer that all parameters affecting safety of `unsafe` blocks are checked or
+prepared close to the `unsafe` code,
+and wherever possible within the same function (thus making the function safe).
+
+### New PRNG Algorithms
+
+The Rand library includes several pseudo-random number generators, and we have
+received several requests to adopt new algorithms into the library. We must
+consider such requests in regards to several things:
+
+-   whether the PRNG is cryptographically secure, and if so, how trustworthy
+    such claims are
+-   statistical quality of output
+-   performance and features of the generator
+-   scope of the project
+-   reception and third-party review of the algorithm
+
+In general, we expect:
+
+-   the author of the algorithm to publish an article of some type (e.g.
+    a scientific article or web page) introducing the new algorithm and
+    discussing its utility, strengths and weaknesses
+-   review of statistical quality and any special features by third parties
+-   good performance in automated test suites like PractRand, TestU01 and
```

Comment:
What is meant by "good performance"? Is this intentionally vague, so it has to be decided on a case-by-case basis?

Reply:
What would you propose then?

I mean obviously "useful performance" depends on what other PRNGs are available with the required features. It seems hard to be specific here.

Reply:
I also depends on the test. In general, it is possible to design a test for any non-CSPRNG where it fails, otherwise it would be a CSPRNG. So for example a strict "no test failure" policy is too simplistic. I think it should be decided on a case-by-case basis.

Reply:
You are aware that not everything is provable, right? For example the ISAAC generator has not been proven either secure or insecure. So yes, one generally has to guess a bit on a case-by-case basis.

Reply:
BigCrush is part of TestU01, some maybe this should just say "good performance in automated test suites like PractRand and BigCrush"

> I also depends on the test. In general, it is possible to design a test for any non-CSPRNG where it fails, otherwise it would be a CSPRNG.

That is a point, we can't say that the PRNG should pass any test one can *create*. PractRand and BigCrush are not made to pick on any specific PRNG algorithm, but to evaluate how good they are for practical scenarios. The results on these test should definitely be a big factor for consideration.

This requirement would make it hard for your favourite Xoroshiro to become part of Rand, and keep most of the old, worse, designs out the door.

Reply:
> PractRand and BigCrush are not made to pick on any specific PRNG algorithm, but to evaluate how good they are for practical scenarios.

This is not accurate. For example, the binary rank test is specifically designed with shift-register generators in mind (see [the original Marsaglia paper](https://ac.els-cdn.com/0024379585901922/1-s2.0-0024379585901922-main.pdf?_tid=7880d45c-cba8-11e7-9088-00000aacb360&acdnat=1510931189_55402fb5880847b5dd7b39325faba973)).

> This requirement would make it hard for your favourite Xoroshiro to become part of Rand, and keep most of the old, worse, designs out the door.

I still think the binary rank failures don't matter, but "my favorite" xoshiro does not have these issues.

Reply:
Can we please stop quibbling about xo*shiro and binary rank failures? The RNGs are perfectly good enough for inclusion in some Rand sub-crate or external crate, and which we use for `SmallRng` is another matter.

This policy is intended as a guideline, not a hard rule.